### PR TITLE
Apply Substitutions to ClaimPatterns When Simplifying

### LINKED ISSUE #2740: Side condition not used in ensures clause evaluation
Tested with commit 07c0fc8e07f3fce05ffdd771cbb2a4765e1ac074

The final configuration on one of the two proof branches contains this (the other branch fails in a similar way):
```
  #Not ( {
    empty
  #Equals
    copy ( empty , I )
  } )
...
#And
  {
    true
  #Equals
    I >Int 0
  }
```

The semantics has a rule for evaluating `copy` that should make this branch succeed:
```
  rule  copy(empty, I:Int) => empty
    requires I >Int 0
```

and the side condition contains `true #Equals I >Int 0` since the start configuration, however, that rule is not applied. After staring at the function evaluation logs for a while, I think it's because the side condition is not used when attempting to apply the rule above.

To reproduce:

tmp.k
```
module TMP
  imports INT
  imports K-EQUAL

  syntax IntList  ::= "empty" | concat(Int, IntList)

  syntax IntList ::= copy(IntList, Int)  [function, functional]

  rule  copy(empty, I:Int) => empty
    requires I >Int 0
  rule  copy(concat(E:Int, Es:IntList), I:Int) => concat(E, copy(Es, I -Int 1))
    requires I >Int 0
  rule  copy(Es:IntList, _:Int) => Es
    [owise]

  syntax KItem ::= iterateList(IntList, Int)

  rule  iterateList(empty, _:Int) => .K  [label(i1)]
  rule  iterateList(concat(_:Int, L:IntList), A:Int) => iterateList(L, A)  [label(i2)]

endmodule
```

tmp-proof.k
```
module TMP-PROOF
  imports TMP

  claim iterateList(Es, I) => .K
    requires I >Int 0
    ensures copy(Es:IntList, I:Int) ==K Es

endmodule
```

### PR DESCRIPTION
**Pull request by JKTKops**
Fixes #2740

The problem appears to arise because substitution information isn't carried into the call to `simplifyRightHandSide` from `simplify'`. This change makes `Kore.Reachability.Claim.simplify'` more closely match `Kore.Rewrite.Rule.Simplify.simplifyClaimRule.worker`.

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [x] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [x] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [x] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/test/Test/Kore/Reachability/OnePathStrategy.hs [hunk 1]
```diff
@@ -810,7 +810,7 @@ test_onePathStrategy =
             left' =
                 Pattern.withCondition
                     Mock.a
-                    (Condition.assign (inject Mock.x) Mock.a)
+                    (Condition.assign (inject Mock.x) Mock.a){substitution = mempty}
```
Comment (ana-pantilie) [inline kore/test/Test/Kore/Reachability/OnePathStrategy.hs:813]: This should just be `left' = Mock.a`. `Condition.assign` will just add that to the substitution, and `substitution = mempty` will delete it.

File: kore/test/Test/Kore/Reachability/Prove.hs [hunk 2]
```diff
@@ -428,7 +428,7 @@ test_proveClaims =
                         )
                     , Pattern.withCondition
                         Mock.b
-                        (Condition.assign (inject Mock.x) Mock.a)
+                        (Condition.assign (inject Mock.x) Mock.a){Condition.substitution = mempty}
```
Comment (ana-pantilie) [inline kore/test/Test/Kore/Reachability/Prove.hs:431]: Same as above.

**Review by ana-pantilie — APPROVED** [review]
I just left a couple of minor comments.

File: kore/test/Test/Kore/Reachability/OnePathStrategy.hs (hunk 1 above)
Comment (JKTKops) [reply kore/test/Test/Kore/Reachability/OnePathStrategy.hs:813]: Some definition-chasing later, I think the `{substitution = mempty}` just undoes whatever `Pattern.withCondition` does. I replaced it with `Pattern.fromTermLike` instead.

**Review by geo2a — APPROVED** [review]

**Pull request merged**